### ISSUE TITLE
Generating MDN annotations in spec output

### ISSUE THREAD

**Issue by sideshowbarker** [w3c/respec#1858]
This is a feature proposal for a new feature to automatically generate MDN annotations in spec output.

For any specific feature in a spec, ReSpec would automatically generate an MDN annotation with:

* a link to any MDN articles that document that feature
* the summary description of that feature from the MDN article (available as tooltip text on hovering over the annoation)
* the (caniuse-like) browser-compatibility support matrix for that feature

A similar feature has been implemented for the HTML spec, and output for that is available here:

https://arcane-cove-57093.herokuapp.com/multipage/

Here’s a ~15-second screen recording showing the annotations as implemented for the HTML spec —

![mdnannos720](https://user-images.githubusercontent.com/194984/46925623-c0d20b80-d067-11e8-85d9-94e70e267135.gif)

Data that can be used to build the annotation content is available at https://github.com/w3c/mdn-spec-links and https://w3c.github.io/mdn-spec-links/#user-content-spec-json-files

For example, data for the Payment Request API spec is available here:

https://w3c.github.io/mdn-spec-links/payment-request.json

**Comment by EvsChen** [comment]
Any updates on this issue? would like to pick this up

**Comment by marcoscaceres** [comment]
feel free to take it. You might be able to integrate with whatever the HTML spec is doing/using.

**Comment by EvsChen** [comment]
Thanks! I looked at the way html spec does this. They look for element `id` in the MDN json to determine whether to insert the MDN annotation or not. Are we going to follow this or use some new attributes?

For me, a prefix id syntax like `<code id="mdn:sessionStorage"></code>` may be more effective in dom parsing.

**Comment by marcoscaceres** [comment]
I'd prefer we do whatever HTML is already doing: that way we can leverage and maybe improve whatever @sideshowbarker already built. Then we don't need to maintain as much code and we get benefit across HTML and all ReSpec specs.

**Comment by EvsChen** [comment]
I think their build tool `whatwg/wattsi` is using pascal. It's kind of hard to "integrate" what they are doing. At best we could follow their procedure.

**Comment by marcoscaceres** [comment]
Ok, I meant more just on using the same HTML IDs/hooks and the JavaScript that is driving the widget/loading the JSON form MDN... assuming the widget above is just being generated with JavaScript. If the integration is happening at using Pascal, then yeah... we will roll our own.

**Comment by sideshowbarker** [comment]
Yeah the main requirement is to make the generation of the MDN annotations fully automatic so that authors don’t have to do anything new. The necessary data is all already available:

* https://w3c.github.io/mdn-spec-links/#user-content-spec-json-files
* https://github.com/w3c/mdn-spec-links

As explained in https://github.com/w3c/mdn-spec-links/blob/master/README.md, the names of the JSON files the match the corresponding spec-shortname keys used in the data provided by https://www.specref.org/

See also the https://github.com/w3c/mdn-spec-links/blob/master/SPECMAP.json file

So functionally, what the behavior in respec needs to do is:

1. Given a particular spec, fetch the corresponding https://w3c.github.io/mdn-spec-links file for it.
2. Load and parse that JSON. The keys are IDs from the spec.
3. For each ID in the spec source, see if there is a matching key in that JSON object, and if so, get all the associated values: the MDN “slugs” associated with that spec ID, along with the summary description, and the browser-compatibility support matrix.
4. Populate an annotation widget using that data, and put it into the spec’s DOM.

**Comment by marcoscaceres** [comment]
@sideshowbarker, the one bit I'm still unclear on is if we need to recreate the MDN widget that's in the GIF above? Or is that something you are adding into the HTML Spec at build time?

**Comment by sideshowbarker** [comment]
The MDN widget that's in the GIF above is something gets added into the HTML spec at build time.

**Comment by marcoscaceres** [comment]
Ok, cool. Thanks for confirming. @EvsChen, guess we have our answer then :) 

@EvsChen, before you dive too deep, maybe describe how you plan to implement this and we can start with a small pull request. This has the potential to be quite a big task, so it might be good to build it up incrementally with a prototype we can refine.

**Comment by EvsChen** [comment]
I made draft plan on this. Each step could fit in a PR.

On a new branch,
1. Add a new plugin called `mdn-annotation` in core. This plugin should fetch and serialize the corresponding json by looking up spec short name in `SPECMAP.json`.
2. Add more functions to the plugin: for all the element (except `sections`?) that has ID in fetched json, add a simple expandable widget with link.
3. Add browser support list in the widget and other UI improvement. May add an option in respect config to control the kind of browsers displayed.

**Comment by marcoscaceres** [comment]
Plan sounds good. Looking forward to seeing what you come up with! Any issues or questions, please reach out. We are always happy to help.

**Comment by EvsChen** [comment]
A little problem when trying to get the mdn spec json file. How should I find the key in [SPECMAP.json](https://github.com/w3c/mdn-spec-links/blob/master/SPECMAP.json) by respecConf?

Using `edDraftURI` is a good idea but I'm not sure if it's required for any spec. Using `shortName` seems not feasible due to the various hostnames, i.e. `w3c.github.io`, `csswg.org` etc.

**Comment by marcoscaceres** [comment]
The URLs in the spec map don't matter I think. You can match on value being the shortName. All short names are unique. So, you can do:

```
// exists? 
Object.entries({"https://w3c.github.io/payment-request/": "payment-request.json"}).some(([,value])=> value === "${conf.shortname}.json")
```

**Comment by marcoscaceres** [comment]
alternative to the above...

**Comment by EvsChen** [comment]
Great, thanks! If `shortName` is sufficient, the url doesn't matter then.

**Comment by EvsChen** [comment]
I've made a PR for the first part of the plugin. PTAL @marcoscaceres. If everything looks good, I could move on.

**Comment by marcoscaceres** [comment]
Closed via #2110
